### PR TITLE
Add WEBGL_multi_draw_instanced_base_vertex_base_instance bindings

### PR DESCRIPTION
**Pull request by shrekshao**
Add WEBGL_multi_draw_instanced_base_vertex_base_instance function bindings to emscripten. Add a test to verify the new functions hook up properly.

~~The shader support for multi_draw and base_vertex/base_instance (`gl_BaseInstance`, `gl_InstanceID`, etc.) seems not working properly yet. Would fix in the next patch.~~

Also fix a small bug in `webgl2_simple_enable_extensions.c` mentioned in #12280 (Thanks @kainino0x for catching the issue)

### CODE REVIEW

**Comment by shrekshao** [comment]
Pinging @juj for review. Thanks!

File: src/library_webgl2.js [hunk 1]
```diff
@@ -976,10 +976,15 @@ var LibraryWebGL2 = {
     _glDrawElements(mode, count, type, indices);
   },
 
+  glDrawArraysInstancedBaseInstanceWEBGL_sig: 'viiiii',
+  glDrawArraysInstancedBaseInstance: 'glDrawArraysInstancedBaseInstanceWEBGL',
+  glDrawArraysInstancedBaseInstanceANGLE: 'glDrawArraysInstancedBaseInstanceWEBGL',
```
Comment (kripken) [inline src/library_webgl2.js:981]: please reorder these so the aliases are after the implementation. that is,
```
X_sig: ..
X: ..

Y: 'X',
Z: 'X',
```

**Review by kripken — COMMENTED** [review]
(CI issues can be fixed by merging in latest trunk, I believe)

**Comment by kainino0x** [comment]
I can also review this (sorry I didn't get to it today)

**Comment by shrekshao** [comment]
There's still some bots failure needs addressing.

**Comment by shrekshao** [comment]
Bot issue is fixed. Shader extension also works properly now.

File: system/include/webgl/webgl2_ext.h [hunk 2]
```diff
@@ -14,3 +14,9 @@ WEBGL_APICALL void GL_APIENTRY emscripten_glDrawArraysInstancedBaseInstanceWEBGL
 WEBGL_APICALL void GL_APIENTRY emscripten_glDrawElementsInstancedBaseVertexBaseInstanceWEBGL(GLenum mode, GLsizei count, GLenum type, const void *offset, GLsizei instanceCount, GLint baseVertex, GLuint baseinstance);
 WEBGL_APICALL void GL_APIENTRY glDrawArraysInstancedBaseInstanceWEBGL(GLenum mode, GLint first, GLsizei count, GLsizei instanceCount, GLuint baseInstance);
 WEBGL_APICALL void GL_APIENTRY glDrawElementsInstancedBaseVertexBaseInstanceWEBGL(GLenum mode, GLsizei count, GLenum type, const void *offset, GLsizei instanceCount, GLint baseVertex, GLuint baseinstance);
+
+// 47. https://www.khronos.org/registry/webgl/extensions/WEBGL_multi_draw_instanced_base_vertex_base_instance/
+WEBGL_APICALL void GL_APIENTRY emscripten_glMultiDrawArraysInstancedBaseInstanceWEBGL(GLenum mode, const GLint* firsts, const GLsizei* counts, const GLsizei* instanceCounts, const GLuint* baseInstances, GLsizei drawcount);
```
Comment (kainino0x) [inline system/include/webgl/webgl2_ext.h:19]: nit: `drawcount` -> `drawCount`

File: system/include/webgl/webgl2_ext.h [hunk 3]
```diff
@@ -14,3 +14,9 @@ WEBGL_APICALL void GL_APIENTRY emscripten_glDrawArraysInstancedBaseInstanceWEBGL
...
+WEBGL_APICALL void GL_APIENTRY emscripten_glMultiDrawElementsInstancedBaseVertexBaseInstanceWEBGL(GLenum mode, const GLsizei* counts, GLenum type, const GLvoid* const* offsets, const GLsizei* instanceCounts, const GLint* baseVertices, const GLuint* baseinstances);
```
Comment (kainino0x) [inline system/include/webgl/webgl2_ext.h:20]: nit: `baseinstances` -> `baseInstances`

File: system/include/webgl/webgl2_ext.h [hunk 4]
```diff
@@ -14,3 +14,9 @@ WEBGL_APICALL void GL_APIENTRY emscripten_glDrawArraysInstancedBaseInstanceWEBGL
...
+WEBGL_APICALL void GL_APIENTRY glMultiDrawArraysInstancedBaseInstanceWEBGL(GLenum mode, const GLint* firsts, const GLsizei* counts, const GLsizei* instanceCounts, const GLuint* baseInstances, GLsizei drawcount);
+WEBGL_APICALL void GL_APIENTRY glMultiDrawElementsInstancedBaseVertexBaseInstanceWEBGL(GLenum mode, const GLsizei* counts, GLenum type, const GLvoid* const* offsets, const GLsizei* instanceCounts, const GLint* baseVertices, const GLuint* baseinstances, GLsizei drawcount);
```
Comment (kainino0x) [inline system/include/webgl/webgl2_ext.h:22]: I think there should probably be a newline at the end of this file.

File: system/include/webgl/webgl2_ext.h (hunk 3 above)
Comment (kainino0x) [inline system/include/webgl/webgl2_ext.h:20]: Is this missing the `drawCount` argument?

File: tests/webgl_draw_base_vertex_base_instance_test.c [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  glShaderSource(shader, 1, &src, NULL);
+  glCompileShader(shader);
+
+  GLint isCompiled = 0;
+  glGetShaderiv(shader, GL_COMPILE_STATUS, &isCompiled);
+  if (!isCompiled)
+  {
+    GLint maxLength = 0;
+    glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &maxLength);
+    char *buf = (char*)malloc(maxLength+1);
```
Comment (kainino0x) [inline tests/webgl_draw_base_vertex_base_instance_test.c:30]: uNit: I don't think `+1` is needed here; GL_INFO_LOG_LENGTH should include space for the null char.

File: tests/webgl_draw_base_vertex_base_instance_test.c [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    glGetShaderInfoLog(shader, maxLength, &maxLength, buf);
```
Comment (kainino0x) [inline tests/webgl_draw_base_vertex_base_instance_test.c:31]: uNit: Should be able to just pass NULL instead of `&maxLength`

**Review by kainino0x — APPROVED** [review]
Overall LGTM, though I would prefer if @juj could take a look as well

**Comment by shrekshao** [comment]
Thanks for the review. Fixed as suggested in the comments.

**Review by kainino0x — APPROVED** [review]

File: src/library_webgl2.js [hunk 7]
```diff
@@ -976,13 +976,18 @@ var LibraryWebGL2 = {
...
   },
 
+  glDrawArraysInstancedBaseInstanceWEBGL_sig: 'viiiii',
   glDrawArraysInstancedBaseInstanceWEBGL: function(mode, first, count, instanceCount, baseInstance) {
     GLctx.dibvbi['drawArraysInstancedBaseInstanceWEBGL'](mode, first, count, instanceCount, baseInstance);
   },
+  glDrawArraysInstancedBaseInstance: 'glDrawArraysInstancedBaseInstanceWEBGL',
+  glDrawArraysInstancedBaseInstanceANGLE: 'glDrawArraysInstancedBaseInstanceWEBGL',
 
+  glDrawElementsInstancedBaseVertexBaseInstanceWEBGL_sig: 'viiiiii',
```
Comment (kripken) [inline src/library_webgl2.js:986]: sig has 6 params, but the function seems to have 7?

Comment (shrekshao) [reply src/library_webgl2.js]: 😰 Fixed. Checked the other three.

**Review by kripken — APPROVED** [review]
lgtm on the non-WebGL side of things.

(https://github.com/emscripten-core/emscripten/pull/12319 will help with argument counts in the future)

**Pull request merged**